### PR TITLE
ICSAAS-307:  Implement Component - Search

### PR DESCRIPTION
**Pull request by jerome-mottet-valtech**

### CODE REVIEW

File: ui.frontend/src/main/webpack/site/js/components/loadMoreButton.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  loadMoreButton.innerText = loadMoreButtonText
+
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
+  loadMoreButton.addEventListener('click', async (event) => {
+    onLoadMoreButtonClick?.()
+
+    event.preventDefault()
+
+    const currentOffset = loadMoreButton.dataset.offset || offset
```
Comment (igorsimovski) [inline ui.frontend/src/main/webpack/site/js/components/loadMoreButton.ts:34]: Probably i've communicated the meaning of "start" param's value incorrectly.
For paginations there are two params used in saas api ("start" and "rows").
start - the result page index in the pagination
rows - the results per page
example 1: start = 0; rows = 10 - will fetch the first 10 result.
example 2: start = 2; rows = 10 - will fetch the results from 21 - 30.
I decided for the AEM search api to use the same param name "start", and rows (configurable in search cmp's dialog) will be applied on BE side. So for the load more the value of start should increase by 1.

File: ui.frontend/src/main/webpack/site/js/components/searchForm.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  searchForm.classList.add('saas-container_form')
+
+  return searchForm
+}
+
+export const triggerSearch = async (
+  searchForm: HTMLFormElement,
+  searchInputElement: HTMLInputElement,
+  searchUrl: string | undefined,
+  searchTabs: string[],
```
Comment (igorsimovski) [inline ui.frontend/src/main/webpack/site/js/components/searchForm.ts:24]: This property has been changed to a array of objects
sample search config json:
`{
  "title" : "Search Cmp",
  "searchFieldPlaceholderText" : "Type search term here...",
  "searchButtonText" : "Search",
  "loadMoreButtonText" : "Load more",
  "searchTabs" : [ {
    "title" : "Search tab",
    "url" : "/content/saas-aem-module/us/en/_jcr_content/root/container/container/search/search-tabs/searchtab.model.json?q=pg",
    ":type" : "saas-aem-module/components/searchtab"
  }, {
    "title" : "ZIP Only",
    "url" : "/content/saas-aem-module/us/en/_jcr_content/root/container/container/search/search-tabs/searchtab_765476375.model.json?q=pg",
    ":type" : "saas-aem-module/components/searchtab"
  } ],
  "autosuggestUrl" : "/content/saas-aem-module/us/en/_jcr_content/root/container/container/search.autocomplete.json",
  "autocompleteTriggerThreshold" : 3,
  ":type" : "saas-aem-module/components/search",
  ":items" : { },
  ":itemsOrder" : [ ]
}`

File: ui.frontend/src/main/webpack/site/js/components/loadMoreButton.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    const result = await fetch(
+      `${tabUrl}&q=${searchValue}&start=${currentOffset || 0}`,
+    )
+
+    loadMoreButton.dataset.offset = `${+currentOffset + offset}`
+
+    const resultJSON = await result.json()
+
+    resultJSON.results.forEach((resultItem: SearchItem) => {
```
Comment (SgiobairOg) [inline ui.frontend/src/main/webpack/site/js/components/loadMoreButton.ts:44]: consider a check here to ensure `resultJSON` has `results` property

**Review by SgiobairOg — APPROVED** [review]

File: ui.frontend/src/main/webpack/site/js/components/searchItem.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+export const buildSearchItem = ({
+  title,
+  description,
+  url,
+}: SearchItem): HTMLDivElement => {
+  const searchItem = document.createElement('div')
+
+  const searchItemTitle = document.createElement('h2')
```
Comment (nhirrle) [inline ui.frontend/src/main/webpack/site/js/components/searchItem.ts:14]: h2 is pretty heavy. h3 sounds more appropriate

File: ui.frontend/src/main/webpack/site/js/components/searchItem.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  description: string
+  url: string
+}
+
+export const buildSearchItem = ({
+  title,
+  description,
+  url,
+}: SearchItem): HTMLDivElement => {
+  const searchItem = document.createElement('div')
```
Comment (nhirrle) [inline ui.frontend/src/main/webpack/site/js/components/searchItem.ts:12]: - [ ] please add classes to those elements so they can be easily styled

File: ui.frontend/src/main/webpack/site/js/components/searchTabs.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  const searchTabName = document.createElement('span')
+  searchTabName.innerHTML = title
+
+  const searchTabNumberOfResults = document.createElement('span')
+  searchTabNumberOfResults.innerHTML = ` (${tabNumberOfResults})`
+
+  searchTab.appendChild(searchTabName)
+  searchTab.appendChild(searchTabNumberOfResults)
+
+  searchTab.addEventListener('click', () => {
```
Comment (nhirrle) [inline ui.frontend/src/main/webpack/site/js/components/searchTabs.ts:38]: wopuld tab-navigationw work as well ?

**Review by nhirrle — APPROVED** [review]
minor comments but code looks good. thanks!

File: ui.frontend/src/main/webpack/site/js/components/loadMoreButton.ts (hunk 1 above)
Comment (jerome-mottet-valtech) [reply ui.frontend/src/main/webpack/site/js/components/loadMoreButton.ts:34]: Fixed. I adapted the offset value when invoking `buildLoadMoreButton` in `searchForm.ts`

**Pull request merged**